src/kbmod/search/geom.h

```diff
@@ -0,0 +1,226 @@
+
+#include <iostream>
+#include <optional>
+#include <utility>  // pair
+#include <array>
+#include <vector>
+#include <assert.h>
+
+#include <Eigen/Core>
+
+#include "common.h"
+//#include "pydocs/geom_docs.h"
+
+namespace indexing {
+
+struct Index {
+    int j;
+    int i;
+
+    const std::string to_string() const {
+        return "Index(" + std::to_string(j) + ", " + std::to_string(i) + ")";
+    }
+
+    const std::string to_yaml() const {
+        return "{j: " + std::to_string(j) + " i: " + std::to_string(i) + "}";
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const Index& rc);
+};
+
+std::ostream& operator<<(std::ostream& os, const Index& rc) {
+    os << rc.to_string();
+    return os;
+}
+
+struct Point {
+    float y;
+    float x;
+
+    const Index to_index() const { return {(int)(floor(y - 0.5f) + 0.5f), (int)(floor(x - 0.5f) + 0.5f)}; }
+
+    const std::string to_string() const {
+        return "Point(" + std::to_string(y) + ", " + std::to_string(x) + ")";
+    }
+
+    const std::string to_yaml() const {
+        return "{x: " + std::to_string(y) + " y: " + std::to_string(x) + "}";
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const Point& rc);
+};
+
+std::ostream& operator<<(std::ostream& os, const Point& rc) {
+    os << rc.to_string();
+    return os;
+}
+
+// A rectangle that also contains it's corner's origin Index with respect to
+// another, larger rectangle.
+struct AnchoredRectangle {
+    Index corner;
+    Index anchor;
+
+    unsigned height;
+    unsigned width;
+
+    const std::string to_string() const {
+        return "AnchoredRectangle(corner:" + corner.to_string() + ", anchor:" + anchor.to_string() +
+               ", height: " + std::to_string(height) + ", width: " + std::to_string(width) + ")";
+    }
+
+    const std::string to_yaml() const {
+        return "{corner: " + corner.to_yaml() + " anchor: " + anchor.to_yaml() +
+               " height: " + std::to_string(height) + " width: " + std::to_string(width) + "}";
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const AnchoredRectangle& rc);
+};
+
+std::ostream& operator<<(std::ostream& os, const AnchoredRectangle& rc) {
+    os << rc.to_string();
+    return os;
+}
+
+// Given an index component `idx_val`, and a radius `r` around it, returns the
+// start and end index components, and length of the range, that fit in the
+// [0, max_range] limits.
+inline std::tuple<int, int, int> centered_range(int idx_val, const int r, const int max_range) {
+    // pin start to the []0, max_range] range
+    int start = std::max(0, idx_val - r);
+    start = std::min(start, max_range);
+
+    // pin end to the [0, max_range]
+    int end = std::max(0, idx_val + r);
+    end = std::min(max_range, idx_val + r);
+
+    // range is inclusive of the first element, and can not be longer than start
+    // minus max range
+    int length = end - start + 1;
+    length = std::min(length, max_range - start);
```

Comment:
I think this is a bit confusing because when start is 0 - we count that as an element, but when start is something like 7 and end is 10, then the length is just `end-start`.

Removing it will break a test in `test_geom.py` - but because that's the new functionality we've just added we can just remove this and fix the tests. I'm seeing something is off (or confusing at least) regarding the range in the tests but left it as is to make progress. Have to revisit, would appreciate a sanity check.